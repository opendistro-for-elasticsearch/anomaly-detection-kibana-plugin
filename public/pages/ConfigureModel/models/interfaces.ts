/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import { AggregationOption } from '../../../models/types';

// Formik values used when creating/editing the model configuration
export interface ModelConfigurationFormikValues {
  featureList: FeaturesFormikValues[];
  categoryFieldEnabled: boolean;
  categoryField: string[];
  shingleSize: number;
}

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string | undefined;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}
