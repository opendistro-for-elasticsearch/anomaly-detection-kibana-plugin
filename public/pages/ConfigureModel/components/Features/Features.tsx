/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiCallOut,
  EuiSpacer,
  EuiText,
  EuiFlexItem,
  EuiFlexGroup,
  EuiButton,
  EuiLink,
  EuiIcon,
} from '@elastic/eui';
import { FieldArray, FieldArrayRenderProps, FormikProps } from 'formik';

import { get } from 'lodash';
import React, { Fragment, useEffect } from 'react';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { Detector } from '../../../../models/interfaces';
import { initialFeatureValue } from '../../utils/helpers';
import { MAX_FEATURE_NUM } from '../../../../utils/constants';
import { FeatureAccordion } from '../FeatureAccordion';

interface FeaturesProps {
  detector: Detector | undefined;
  formikProps: FormikProps<any>;
}

export function Features(props: FeaturesProps) {
  // If the features list is empty: push a default initial one
  useEffect(() => {
    if (get(props, 'formikProps.values.featureList', []).length === 0) {
      props.formikProps.setFieldValue('featureList', [initialFeatureValue()]);
      props.formikProps.setFieldTouched('featureList', false);
    }
  }, [props.formikProps.values.featureList]);

  return (
    <ContentPanel
      title="Features"
      titleSize="s"
      subTitle={
        <EuiText className="content-panel-subTitle">
          Specify an index field that you want to find anomalies for by defining
          features. You can add up to 5 features.{' '}
          <EuiLink
            href="https://opendistro.github.io/for-elasticsearch-docs/docs/ad/"
            target="_blank"
          >
            Learn more <EuiIcon size="s" type="popout" />
          </EuiLink>
        </EuiText>
      }
    >
      <EuiFlexGroup direction="column" style={{ margin: '0px' }}>
        <FieldArray name="featureList" validateOnChange={true}>
          {({ push, remove, form: { values } }: FieldArrayRenderProps) => {
            return (
              <Fragment>
                {get(props.detector, 'indices.0', '').includes(':') ? (
                  <div>
                    <EuiCallOut
                      title="This detector is using a remote cluster index, so you need to manually input the field."
                      color="warning"
                      iconType="alert"
                    />
                    <EuiSpacer size="m" />
                  </div>
                ) : null}
                {values.featureList.map((feature: any, index: number) => (
                  <FeatureAccordion
                    onDelete={() => {
                      remove(index);
                    }}
                    index={index}
                    feature={feature}
                    handleChange={props.formikProps.handleChange}
                  />
                ))}
                <EuiFlexGroup
                  alignItems="center"
                  style={{ padding: '12px 0px' }}
                >
                  <EuiFlexItem grow={false}>
                    <EuiButton
                      data-test-subj="addFeature"
                      isDisabled={values.featureList.length >= MAX_FEATURE_NUM}
                      onClick={() => {
                        push(initialFeatureValue());
                      }}
                    >
                      Add another feature
                    </EuiButton>
                    <EuiText className="content-panel-subTitle">
                      <p>
                        You can add up to{' '}
                        {Math.max(
                          MAX_FEATURE_NUM - values.featureList.length,
                          0
                        )}{' '}
                        more features.
                      </p>
                    </EuiText>
                  </EuiFlexItem>
                </EuiFlexGroup>
              </Fragment>
            );
          }}
        </FieldArray>
      </EuiFlexGroup>
    </ContentPanel>
  );
}
