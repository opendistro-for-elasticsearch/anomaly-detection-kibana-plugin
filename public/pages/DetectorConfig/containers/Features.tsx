/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
import React, { Component } from 'react';
import {
  EuiBasicTable,
  EuiText,
  EuiLink,
  EuiIcon,
  EuiButton,
  EuiEmptyPrompt,
} from '@elastic/eui';
import {
  Detector,
  FEATURE_TYPE,
  FeatureAttributes,
} from '../../../models/interfaces';
import { get, sortBy } from 'lodash';
import { PLUGIN_NAME } from '../../../utils/constants';
import ContentPanel from '../../../components/ContentPanel/ContentPanel';
import { CodeModal } from '../components/CodeModal/CodeModal';

interface FeaturesProps {
  detectorId: string;
  detector: Detector;
  onEditFeatures(): void;
}

interface FeaturesState {
  showCodeModel: boolean[];
  sortField: string;
  sortDirection: 'asc' | 'desc';
}

export class Features extends Component<FeaturesProps, FeaturesState> {
  constructor(props: FeaturesProps) {
    super(props);
    this.state = {
      showCodeModel: get(props.detector, 'featureAttributes', []).map(
        () => false
      ),
      sortField: 'name',
      sortDirection: 'asc',
    };
  }

  private closeModal(index: number) {
    const cloneShowCodeModal = [...this.state.showCodeModel];
    cloneShowCodeModal[index] = false;
    this.setState({
      showCodeModel: cloneShowCodeModal,
    });
  }

  private showModal(index: number) {
    const cloneShowCodeModal = [...this.state.showCodeModel];
    cloneShowCodeModal[index] = true;
    this.setState({
      showCodeModel: cloneShowCodeModal,
    });
  }

  private getModalVisibilityChange = (index: number) => {
    return this.state.showCodeModel[index];
  };

  private handleTableChange = (props: any) => {
    this.setState({
      sortField: props.sort.field,
      sortDirection: props.sort.direction,
    });
  };

  private sortedItems(items: Array<any>) {
    let sorted = sortBy(items, this.state.sortField);
    if (this.state.sortDirection == 'desc') {
      sorted = sorted.reverse();
    }
    return sorted;
  }

  public render() {
    const featureAttributes = get(this.props.detector, 'featureAttributes', []);

    const sorting = {
      sort: {
        field: this.state.sortField,
        direction: this.state.sortDirection,
      },
    };

    const items = featureAttributes.map(
      (feature: FeatureAttributes, index: number) => ({
        name: feature.featureName,
        definition: index,
        state: feature.featureEnabled ? 'Enabled' : 'Disabled',
      })
    );

    const sortedItems = this.sortedItems(items);

    const columns = [
      {
        field: 'name',
        name: 'Feature name',
        sortable: true,
      },
      {
        field: 'definition',
        name: 'Feature definition',
        render: (featureIndex: number) => {
          const feature = featureAttributes[featureIndex];

          const metaData = get(
            this.props.detector,
            `uiMetadata.features.${feature.featureName}`,
            {}
          );

          if (
            Object.keys(metaData).length === 0 ||
            metaData.featureType == FEATURE_TYPE.CUSTOM
          ) {
            return (
              <div>
                <p>
                  Custom expression:{' '}
                  <EuiLink
                    data-test-subj={`viewFeature-${featureIndex}`}
                    onClick={() => this.showModal(featureIndex)}
                  >
                    View code
                  </EuiLink>
                </p>

                {!this.getModalVisibilityChange(featureIndex) ? null : (
                  <CodeModal
                    code={JSON.stringify(feature.aggregationQuery, null, 4)}
                    title={feature.featureName}
                    subtitle="Custom expression"
                    closeModal={() => this.closeModal(featureIndex)}
                    getModalVisibilityChange={() =>
                      this.getModalVisibilityChange(featureIndex)
                    }
                  />
                )}
              </div>
            );
          } else {
            return (
              <div>
                <p>Field: {metaData.aggregationOf || ''}</p>
                <p>Aggregation method: {metaData.aggregationBy || ''}</p>
              </div>
            );
          }
        },
      },
      {
        field: 'state',
        name: 'State',
      },
    ];

    const getCellProps = () => {
      return {
        textOnly: true,
      };
    };

    const featureNum = Object.keys(featureAttributes).length;

    return (
      <ContentPanel
        title={`Features (${featureNum})`}
        titleSize="s"
        subTitle={
          <EuiText className="anomaly-distribution-subtitle">
            <p>
              Specify index fields that you want to find anomalies for by
              defining features. A detector can discover anomalies for up to 5
              features. Once you define the features, you can preview your
              anomalies from a sample feature output.{' '}
              <EuiLink
                href="https://opendistro.github.io/for-elasticsearch-docs/docs/ad/"
                target="_blank"
              >
                Learn more &nbsp;
                <EuiIcon size="s" type="popout" />
              </EuiLink>
            </p>
          </EuiText>
        }
        actions={[
          <EuiButton onClick={this.props.onEditFeatures}>Edit</EuiButton>,
        ]}
      >
        {featureNum == 0 ? (
          <EuiEmptyPrompt
            title={
              <span className="emptyFeatureTitle">
                Features are required to run a detector
              </span>
            }
            body={
              <EuiText className="emptyFeatureBody">
                Specify index fields that you want to find anomalies for by
                defining features. Once you define the features, you can preview
                your anomalies from a sample feature output.
              </EuiText>
            }
            actions={[
              <EuiButton
                data-test-subj="createButton"
                href={`${PLUGIN_NAME}#/detectors/${this.props.detectorId}/features`}
                fill
              >
                Add feature
              </EuiButton>,
            ]}
          />
        ) : (
          <EuiBasicTable
            items={sortedItems}
            columns={columns}
            cellProps={getCellProps}
            sorting={sorting}
            onChange={this.handleTableChange}
          />
        )}
      </ContentPanel>
    );
  }
}
