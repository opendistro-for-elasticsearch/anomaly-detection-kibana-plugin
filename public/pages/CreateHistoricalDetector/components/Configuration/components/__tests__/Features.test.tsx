/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import {
  HashRouter as Router,
  RouteComponentProps,
  Route,
  Switch,
} from 'react-router-dom';
import { render } from '@testing-library/react';
import { Features } from '../Features/Features';
import configureStore from '../../../../../../redux/configureStore';
import {
  httpClientMock,
  coreServicesMock,
} from '../../../../../../../test/mocks';
import { CoreServicesContext } from '../../../../../../components/CoreServices/CoreServices';
import { Formik } from 'formik';
import { FEATURE_TYPE } from '../../../../../../models/interfaces';

const TITLE_TEXT = 'Features';
const DEFAULT_FEATURE_TEXT = 'Add feature';
const TEST_FEATURE_NAME = 'test-feature';
const TEST_FEATURE = {
  featureId: 'test-feature-id',
  featureName: 'test-feature',
  featureType: FEATURE_TYPE.SIMPLE,
  aggregationBy: 'sum',
};

const renderWithRouter = (
  detector: {} = {},
  featureArray: [] = [],
  isLoading: boolean = false
) => ({
  ...render(
    <Provider store={configureStore(httpClientMock)}>
      <Router>
        <Switch>
          <Route
            render={(props: RouteComponentProps) => (
              <CoreServicesContext.Provider value={coreServicesMock}>
                <Formik
                  initialValues={{ featureList: featureArray }}
                  onSubmit={jest.fn()}
                >
                  {() => (
                    <div>
                      <Features
                        detector={detector}
                        formikProps={{
                          values: { featureList: featureArray },
                          setFieldValue: jest.fn(),
                          setFieldTouched: jest.fn(),
                        }}
                        isLoading={isLoading}
                        {...props}
                      />
                    </div>
                  )}
                </Formik>
              </CoreServicesContext.Provider>
            )}
          />
        </Switch>
      </Router>
    </Provider>
  ),
});

describe('<Features /> spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders the component in loading state', () => {
    const { container, getByText, queryByText } = renderWithRouter(
      {},
      [],
      true
    );
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
    expect(queryByText(DEFAULT_FEATURE_TEXT)).toBeNull();
  });
  test('renders the component with no given features', () => {
    const { container, getByText } = renderWithRouter({}, [], false);
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
  });
  test('renders the component with one feature', () => {
    console.error = jest.fn();
    const { container, getByText } = renderWithRouter(
      {},
      [TEST_FEATURE],
      false
    );
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
    getByText(TEST_FEATURE_NAME);
    getByText('sum()');
  });
});
