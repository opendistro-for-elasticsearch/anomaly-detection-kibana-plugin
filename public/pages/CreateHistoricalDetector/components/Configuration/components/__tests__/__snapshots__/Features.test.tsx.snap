// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Features /> spec renders the component in loading state 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Features
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--medium content-panel-subTitle"
            >
              Specify an index field that you want to find anomalies for by defining features. You can add up to 5 features.
               
              <a
                class="euiLink euiLink--primary"
                href="https://opendistro.github.io/for-elasticsearch-docs/docs/ad/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Learn more 
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--small euiIcon-isLoading"
                  focusable="false"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
        style="margin: 0px;"
      >
        <span
          class="euiLoadingSpinner euiLoadingSpinner--large"
          style="margin-left: 4px; margin-right: 8px; margin-bottom: 8px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Features /> spec renders the component with no given features 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Features
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--medium content-panel-subTitle"
            >
              Specify an index field that you want to find anomalies for by defining features. You can add up to 5 features.
               
              <a
                class="euiLink euiLink--primary"
                href="https://opendistro.github.io/for-elasticsearch-docs/docs/ad/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Learn more 
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--small"
                  focusable="false"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M13 8.5a.5.5 0 111 0V12a2 2 0 01-2 2H4a2 2 0 01-2-2V4a2 2 0 012-2h3.5a.5.5 0 010 1H4a1 1 0 00-1 1v8a1 1 0 001 1h8a1 1 0 001-1V8.5zm-5.12.339a.5.5 0 11-.706-.707L13.305 2H10.5a.5.5 0 110-1H14a1 1 0 011 1v3.5a.5.5 0 11-1 0V2.72L7.88 8.838z"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
        style="margin: 0px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 12px 24px;"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <button
              class="euiButton euiButton--primary"
              data-test-subj="addFeature"
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Add feature
                </span>
              </span>
            </button>
            <div
              class="euiText euiText--medium content-panel-subTitle"
            >
              <p>
                You can add up to
                 
                5
                 
                more features.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Features /> spec renders the component with one feature 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Features
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--medium content-panel-subTitle"
            >
              Specify an index field that you want to find anomalies for by defining features. You can add up to 5 features.
               
              <a
                class="euiLink euiLink--primary"
                href="https://opendistro.github.io/for-elasticsearch-docs/docs/ad/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Learn more 
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--small"
                  focusable="false"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M13 8.5a.5.5 0 111 0V12a2 2 0 01-2 2H4a2 2 0 01-2-2V4a2 2 0 012-2h3.5a.5.5 0 010 1H4a1 1 0 00-1 1v8a1 1 0 001 1h8a1 1 0 001-1V8.5zm-5.12.339a.5.5 0 11-.706-.707L13.305 2H10.5a.5.5 0 110-1H14a1 1 0 011 1v3.5a.5.5 0 11-1 0V2.72L7.88 8.838z"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
        style="margin: 0px;"
      >
        <div
          class="euiAccordion euiAccordion__noTopBorder"
        >
          <div
            class="euiAccordion__triggerWrapper"
          >
            <button
              aria-controls="featureList.0"
              aria-expanded="false"
              class="euiAccordion__button euiAccordionForm__noTopPaddingButton"
              id="random_html_id"
              type="button"
            >
              <span
                class="euiAccordion__iconWrapper"
              >
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--medium euiIcon-isLoading euiAccordion__icon"
                  focusable="false"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </span>
              <span
                class="euiIEFlexWrapFix"
              >
                <div
                  id="featureAccordionHeaders.0"
                >
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
                  >
                    <div
                      class="euiFlexItem"
                    >
                      <h5
                        class="euiTitle euiTitle--xsmall euiAccordionForm__title"
                      >
                        test-feature
                      </h5>
                    </div>
                  </div>
                  <span
                    class="content-panel-subTitle"
                    style="padding-right: 20px;"
                  >
                    Field: 
                  </span>
                  <span
                    class="content-panel-subTitle"
                    style="padding-right: 20px;"
                  >
                    Aggregation method: 
                    sum
                  </span>
                  <span
                    class="content-panel-subTitle"
                  >
                    State: 
                    Disabled
                  </span>
                </div>
              </span>
            </button>
            <div
              class="euiAccordion__optionalAction"
            >
              <button
                class="euiButton euiButton--danger euiButton--small euiButton-isDisabled"
                disabled=""
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    Delete
                  </span>
                </span>
              </button>
            </div>
          </div>
          <div
            class="euiAccordion__childWrapper"
            id="featureList.0"
          >
            <div>
              <div
                class="euiAccordion__padding--l"
              >
                <div
                  class="euiFormRow"
                  id="random_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="random_html_id"
                    >
                      Feature name
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFormControlLayout"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <input
                          aria-describedby="random_html_id-help"
                          class="euiFieldText"
                          id="random_html_id"
                          name="featureList.0.featureName"
                          placeholder="Enter feature name"
                          type="text"
                          value="test-feature"
                        />
                      </div>
                    </div>
                    <div
                      class="euiFormHelpText euiFormRow__text"
                      id="random_html_id-help"
                    >
                      Enter a descriptive name. The name must be unique within this detector. Feature name must contain 1-64 characters. Valid characters are a-z, A-Z, 0-9, -(hyphen) and _(underscore).
                    </div>
                  </div>
                </div>
                <div
                  class="euiFormRow"
                  id="random_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="random_html_id"
                    >
                      Feature state
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiCheckbox"
                    >
                      <input
                        class="euiCheckbox__input"
                        id="random_html_id"
                        name="featureList.0.featureEnabled"
                        type="checkbox"
                        value=""
                      />
                      <div
                        class="euiCheckbox__square"
                      />
                      <label
                        class="euiCheckbox__label"
                        for="random_html_id"
                      >
                        Enable feature
                      </label>
                    </div>
                  </div>
                </div>
                <div
                  class="euiFormRow"
                  id="random_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="random_html_id"
                    >
                      Find anomalies based on
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFormControlLayout"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <select
                          class="euiSelect"
                          id="random_html_id"
                          name="featureList.0.featureType"
                        >
                          <option
                            value="simple_aggs"
                          >
                            Field value
                          </option>
                          <option
                            value="custom_aggs"
                          >
                            Custom expression
                          </option>
                        </select>
                        <div
                          class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                        >
                          <span
                            class="euiFormControlLayoutCustomIcon"
                          >
                            <svg
                              aria-hidden="true"
                              class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                              focusable="false"
                              height="16"
                              role="img"
                              viewBox="0 0 16 16"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            />
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="euiFormRow"
                  id="random_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="random_html_id"
                    >
                      Aggregation method
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFormControlLayout"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <select
                          aria-describedby="random_html_id-help"
                          class="euiSelect"
                          data-test-subj="aggregationType"
                          id="random_html_id"
                          name="featureList.0.aggregationBy"
                        >
                          <option
                            value="avg"
                          >
                            average()
                          </option>
                          <option
                            value="value_count"
                          >
                            count()
                          </option>
                          <option
                            value="sum"
                          >
                            sum()
                          </option>
                          <option
                            value="min"
                          >
                            min()
                          </option>
                          <option
                            value="max"
                          >
                            max()
                          </option>
                        </select>
                        <div
                          class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                        >
                          <span
                            class="euiFormControlLayoutCustomIcon"
                          >
                            <svg
                              aria-hidden="true"
                              class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                              focusable="false"
                              height="16"
                              role="img"
                              viewBox="0 0 16 16"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            />
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="euiFormHelpText euiFormRow__text"
                      id="random_html_id-help"
                    >
                      The aggregation method determines what constitutes an anomaly. For example, if you choose min(), the detector focuses on finding anomalies based on the minimum values of your feature.
                    </div>
                  </div>
                </div>
                <div
                  class="euiFormRow"
                  id="random_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="random_html_id"
                    >
                      Field
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      class="euiComboBox"
                      name="featureList.0.aggregationOf"
                      role="combobox"
                    >
                      <div
                        class="euiFormControlLayout"
                      >
                        <div
                          class="euiFormControlLayout__childrenWrapper"
                        >
                          <div
                            class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                            data-test-subj="comboBoxInput"
                            tabindex="-1"
                          >
                            <p
                              class="euiComboBoxPlaceholder"
                            >
                              Select field
                            </p>
                            <div
                              class="euiComboBox__input"
                              style="font-size: 14px; display: inline-block;"
                            >
                              <input
                                aria-controls=""
                                data-test-subj="comboBoxSearchInput"
                                id="random_html_id"
                                role="textbox"
                                style="box-sizing: content-box; width: 2px;"
                                value=""
                              />
                              <div
                                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                              />
                            </div>
                          </div>
                          <div
                            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                          >
                            <button
                              aria-label="Open list of options"
                              class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                              data-test-subj="comboBoxToggleListButton"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                focusable="false"
                                height="16"
                                role="img"
                                viewBox="0 0 16 16"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              />
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 12px 24px;"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <button
              class="euiButton euiButton--primary"
              data-test-subj="addFeature"
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Add feature
                </span>
              </span>
            </button>
            <div
              class="euiText euiText--medium content-panel-subTitle"
            >
              <p>
                You can add up to
                 
                4
                 
                more features.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
